### PR TITLE
auth: Split SAML connector code to own type

### PR DESCRIPTION
**Pull request by camscale**
Create a new type - `SAMLAuthService` - and move the implementation of
the SAML connector from the auth Service to this new type. This
decoupling will make it easier to move out of tree in a later commit.

This also lays the basic groundwork for running without the SAML
connector at all, returning a NotImplemented error for all the SAML
logic if the SAMLAuthService is not plugged in.

A number of names in the `auth` package have been exported to facilitate
this. While technically not necessary for this PR as the new type lives
in this `auth` package, these changes are to make the move of the SAML
connector code to the enterprise repository as simple as possible and be
almost a cut and paste. This allows the changes here to be reviewed more
easily, rather than requiring PRs be compared across two repositories.

Issue: https://github.com/gravitational/teleport.e/issues/525

---

NOTE: This will be easier to review going commit by commit. Each commit
is reasonably small and self-contained.

https://github.com/gravitational/teleport.e/pull/597 and https://github.com/gravitational/teleport/pull/18285 are follow-on PRs that move the refactored SAML connector to the enterprise repo and remove it from the OSS repo.

https://github.com/gravitational/teleport.e/pull/606 and https://github.com/gravitational/teleport/pull/18419 are follow-on PRs that move the SAML auth web handlers to the enterprise repo and remove it from the OSS repo.

### CODE REVIEW

File: lib/auth/sso_diag_context.go [hunk 1]
```diff
@@ -22,33 +22,48 @@
...
+	return f(ctx, authKind, authRequestID, entry)
+}
+
+// WriteToBackend saves the accumulated SSO diagnostic information to the backend.
+func (c *SSODiagContext) WriteToBackend(ctx context.Context) {
+	if c.Info.TestFlow {
+		err := c.DiagService.CreateSSODiagnosticInfo(ctx, c.AuthKind, c.RequestID, c.Info)
 		if err != nil {
-			log.WithError(err).WithField("requestID", c.requestID).Warn("failed to write SSO diag info data")
+			log.WithError(err).WithField("requestID", c.RequestID).Warn("failed to write SSO diag info data")
```
[the remark this answers is not in the document]
Comment (camscale) [reply lib/auth/sso_diag_context.go:57]: This seems a little bogus. This code is providing structured data to logrus. It should be sanitizing it before outputting it and likely would for JSON output. If this does need to be sanitized, what sanitization do we need to perform?

**Comment by zmb3** [comment]
Haven't reviewed yet, but I wonder if we've gone so far to make a separate logical "service" for managing SAML connectors, maybe we should do the same at the gRPC level to start to improve the bloated auth.service.

**Comment by camscale** [comment]
> Haven't reviewed yet, but I wonder if we've gone so far to make a separate logical "service" for managing SAML connectors, maybe we should do the same at the gRPC level to start to improve the bloated auth.service.

@zmb3 We may get there, but one step at a time. I would like to see the auth service split up - it's a bit of a beast to work with, but that is too much for this PR.

**Comment by camscale** [comment]
@tigrato @zmb3 Friendly ping

**Review by tigrato — APPROVED** [review]

File: lib/auth/auth.go [hunk 2]
```diff
@@ -474,6 +480,13 @@ type Server struct {
 	loadAllCAs bool
 }
 
+// RegisterSAMLService registers ss as the SAMLService that provides the SAML
+// connector implementation. If a SAMLService has already been registered, this
+// will override the previous registration.
+func (a *Server) RegisterSAMLService(ss SAMLService) {
```
Comment (r0mant) [inline lib/auth/auth.go:486]: Nit: I think `SetSAMLService` would be more in line with how we usually name methods like this.

File: lib/auth/saml.go [hunk 3]
```diff
@@ -69,11 +165,11 @@ func (a *Server) UpsertSAMLConnector(ctx context.Context, connector types.SAMLCo
...
 
 // DeleteSAMLConnector deletes a SAML connector by name.
-func (a *Server) DeleteSAMLConnector(ctx context.Context, connectorName string) error {
-	if err := a.Services.DeleteSAMLConnector(ctx, connectorName); err != nil {
+func (sas *SAMLAuthService) DeleteSAMLConnector(ctx context.Context, connectorName string) error {
+	if err := sas.auth.Services.DeleteSAMLConnector(ctx, connectorName); err != nil {
 		return trace.Wrap(err)
 	}
-	if err := a.emitter.EmitAuditEvent(ctx, &apievents.OIDCConnectorDelete{
+	if err := sas.emitter.EmitAuditEvent(ctx, &apievents.SAMLConnectorDelete{
```
Comment (r0mant) [inline lib/auth/saml.go:172]: Nice typo catch.

**Review by r0mant — APPROVED** [review]

File: lib/auth/auth.go (hunk 2 above)
Comment (camscale) [reply lib/auth/auth.go]: Done: 22d628c80582a5f099e779a734062c5eb3575295

**Pull request merged**